internals/plan/plan.go

```diff
@@ -1105,3 +1116,75 @@ func ReadDir(dir string) (*Plan, error) {
 	}
 	return plan, err
 }
+
+// MergeServiceContext merges the overrides on top of the service context
+// specified by serviceName, returning a new ContextOptions value. If
+// serviceName is "" (context not specified), return overrides directly.
+func MergeServiceContext(p *Plan, serviceName string, overrides ContextOptions) (ContextOptions, error) {
+	if serviceName == "" {
+		return overrides, nil
+	}
+	var service *Service
+	for _, s := range p.Services {
+		if s.Name == serviceName {
+			service = s
+			break
+		}
+	}
+	if service == nil {
+		return ContextOptions{}, fmt.Errorf("context service %q not found", serviceName)
+	}
+
+	// Start with the config values from the context service.
+	merged := ContextOptions{
+		Environment: make(map[string]string),
+	}
+	for k, v := range service.Environment {
+		merged.Environment[k] = v
+	}
+	if service.UserID != nil {
+		copied := *service.UserID
+		merged.UserID = &copied
+	}
+	merged.User = service.User
+	if service.GroupID != nil {
+		copied := *service.GroupID
+		merged.GroupID = &copied
+	}
+	merged.Group = service.Group
+	merged.WorkingDir = service.WorkingDir
+
+	// Merge in fields from the overrides, if set.
+	for k, v := range overrides.Environment {
+		merged.Environment[k] = v
+	}
+	if overrides.UserID != nil {
```

Comment:
A couple of functions would turn all of these into oneliners and make these read like a nice sequence:
```
func copyIntPtr(v *int) *int
func copyStrPtr(v *string) *string
```

Reply:
Good call for `copyIntPtr` -- done. We don't need `copyStrPtr` as the strings aren't actually pointers (empty string means not set). Also note that we still need to have the `if overrides.UserID != nil` tests, because you only want to set the merged version at all if the override is non-nil. However, the helper function still saves an awkward line of code in each place, so I've added it.